Joken 2.0 initial commit

victorolinasc opened the pull request:
This is the long talked about Joken 2.0.0 PR. This is a different API and therefore we would like more opinions on the current status.

Joken 2.0 brings:

- A less confusing API: dynamic vs static claims confusion.
- Portable and encapsulated configuration: we recommend the module approach here where you can add your functionality at will.
- Better error handling and messages following the approach by [Michal Muskala](http://michal.muskala.eu/2017/02/10/error-handling-in-elixir-libraries.html).
- Signers can be configured through mix.config and so be different for each environment.
- Big performance impact measured against current implementation and other JWT libraries out there.
- Property tested (just a bit at least).
- Hooks mechanism for extensibility (logging and so on).
- Better docs with some guides on common questions: signer configuration, public/private key confusion and so on.
- Time mock for testing time relative things.
- Jason module for JOSE.

This is an early on implementation with some things still missing. Here is a by all means not complete list of TODOs:

- [ ] Discuss which options we should support in `Joken.Claim`. These could include: optinal, error message and so on.
- [x] Discuss having a better hooks plugging mechanism. It should be possible to add more than one hook without overriding functions. We could append hooks at compile time and just call one by one during execution time.
- [x] Check JWKS with dynamic keys (should look at other APIs... we could use Guardian's implementation and update our key configuration guide).
- [ ] Preload all keys in application bootstrap so that we can fail early and have a map with already parsed keys.
- [x] Another full pass on documentation.
- [x] Typespec everything with better typespecs (internal types for better documentation).
- [ ] Logo?
- [ ] Include HTML reports for benchmarks (Benchee HTML).
- [ ] Create 1.x branch.
- [ ] Clean CHANGELOG pointing to 1.x branch if the user wants to check it out.
- [x] Reach at least 90% coverage.
- [ ] See if the optimizations for [`Poison`](https://github.com/potatosalad/erlang-jose/blob/master/lib/jose/poison/lexical_encoder.ex) json module can be used with `Jason`. This is certainly a hot path for `JOSE`.
- [ ] Discuss putting the override of base64url on a library (**IMPORTANT!!!**).
- [ ] Discuss Joken.Plug (should we keep it here or document a better way of handling this like Mithril??).
- [ ] Discuss an API for parsing token into known struct (maybe implemented as a hook?).
- [ ] Discuss having a repository for benchmarking all other types of token libraries (Guardian, AuthSecret and so on)
- [ ] Discuss making samples in another repository.
- [ ] Discuss having an organization in Github for managing all that.
- [ ] Missing guides:
  - [ ] Differences with Phoenix.Token
  - [x] Differences with using raw JOSE
  - [x] Key configuration (WIP)
  - [x] Testing

This is what I've came up with up to now. I'm sure more things will arise while doing the remaining tasks.

Cheers,

bryanjos commented on lib/joken/application.ex:
```diff
@@ -0,0 +1,32 @@
+  Application behaviour for Joken. Used to bootstrap some key options.
+  """
+
+  def start(_type, _args) do
```
Since there is an application, is it possible to pass in configuration here instead of using the config file?

victorolinasc replied:
Not sure what you mean here. Do you mean passing configuration through release files?

bryanjos commented on lib/joken/current_time.ex:
```diff
@@ -0,0 +1,13 @@
+defmodule Joken.CurrentTime.OS do
```
Can we get rid of this now and use Elixir's built in functions? https://hexdocs.pm/elixir/DateTime.html#to_unix/2. Only they are equivalent

victorolinasc replied:
Done!

victorolinasc commented:
I thought a little about the hooks mechanism and here are things we could support:

1. Several hooks: e.g. one for auditing another for storing the result in a session)
2 Halting mechanism: like Plug halt
3. Bypassing default logic: currently, all hooks execute the mains verify and so on...

For number 1, the issue is that currently it is not pluggable... you have to override. We could provide a hooks function that could return a list of `Joken.Hook` implementations and apply all hooks in order. It could have a macro for building up this list (but we could avoid the macro way too).

For number 2, the easiest way to implement it would be just a contract of return from all hook functions. We could have something like `:halt`, `{:ok, <result of the normal call>}` and `{:skip_call, <result of the NEXT call>}`. I don't like it, but it is easy to add...

Number 3 is also covered with the return trick.

@bryanjos wdyt?

victorolinasc commented:
@ybod Since you've proposed the options to validation for version 1, what do you think we should include there? 

Also, I'll close your PR because we want to include it in version 2 which has a totally different API (sorry for that). Please, contribute to the discussion :)

Thanks!

victorolinasc commented:
@bryanjos I guess this needs another look (somewhat a big update with tests and coverage :)

Core chores we still need before releasing an alpha would be typespec and document everything. 

I thought about providing a DebugHook in the library just as a convenience during development. It might help setting up things during the migration from 1 to 2.

victorolinasc commented:
I've given another pass on docs and typespecs. The last missing thing for before a 2.0-alpha1 release is options to hooks.

A very usable hook would be one to fetch and cache JWKS keys from OpenID Connect servers ([here](https://connect2id.com/products/server/docs/api/jwk-set) and [here](https://auth0.com/docs/jwks) have some good examples). So, we would pass the JWKS endpoint and it would fetch and cache it for future token validations. I'll try adding this on a separate library (could be joken_jwks). The implementation is pretty simple: on `before_sign` and `before_verify` we check if we have a cached signer. If not, we simply fetch and parse it from the JWKS endpoint and pass the signer as an argument. There could be further configuration options like only using it for verifying, matching on key id from the JWT header and so on.

Anyway, this is on 96% coverage currently witch is nice and helped me uncover some nasty details. I am pretty confident on releasing this after the hook options feature.

bryanjos commented on .travis.yml:
```diff
@@ -2,24 +2,12 @@ sudo: false # to use faster container based build environment
-  - 1.3.2
-  - 1.4.0
+  - 1.5.3
+  - 1.6.4
```
Can probably update this to 1.6.6 now

bryanjos commented on .travis.yml:
```diff
@@ -2,24 +2,12 @@ sudo: false # to use faster container based build environment
-env:
-  - CC=clang-3.7 CXX=clang++-3.7 ARCHFLAGS=-Wgcc-compat
+  - 19.3
+  - 20.3
```
Can add 21.0 to matrix as well

bryanjos commented on guides/introduction.md:
````diff
@@ -0,0 +1,74 @@
+
+``` elixir
+# config/dev.exs 
+config :joken, default_signer: "secret"
````
Given the recent talk about configuration, do we need to take anything in consideration when defining configs here? https://hexdocs.pm/elixir/master/library-guidelines.html#avoid-application-configuration

victorolinasc replied:
My opinion is that with the latest changes it is already following the guide. I mean: we accept a signer as a parameter already in every possible function call we'd need a signer. So, currently, it is there more as convenience than anything.

We can maybe think about some initialization setup like phoenix and ecto do since we already have an application implementation. I'd add it to 2.1 though for not blocking this further.

I'll work on hook options then! :)

victorolinasc commented:
I've changed how hooks work because of my JWKs strategy: now after hooks receive a result (that can either be `:ok`, `:error` or `{:error, reason}`. That means when we want to halt execution we must return `:halt`. This is because I want to give a chance for the hook to act on an error (like the signer is no longer valid).

Also, I've made `Joken.CurrentTime` a behaviour so that people using `mox` or something similar can mock it easily.

I've also included multi claim validation which is useful when you receive a payload with: `{"access_token": "huge_token", "expires_in": 6666666}` or something like this. 

I've added dialyzer and credo to travis and I'd like your opinion on adding git hooks for formatting and running tests. @bryanjos wdyt? There are a couple of libraries now that make this very simple.

What is missing from my point of view: 
- Discussion on options for claims. The field is already there on the struct so I wouldn't wait for this to release some beta. A minor update wouldn't break anyone who depends on 2.0;
- Some proper way for a hook to retry last execution. Like refetching the key and trying again. This would allow for a JWKS hook to try refetching the signer on error;
- Migration guide from 1.x. I might work on that later this week or next week;
- The preloading of keys is an important usability aspect IMHO. I might work on that too. This might be the last piece that may break before 2.0 beta. I'll namespace the keys on the configuration to make it easier to implement this.

So, what I think we should not implement or hold anymore:

- Forget about docs comparing against Phoenix.Token. IMHO this is quite apparent already on the current docs;
- The samples, benchmarks and etc we can setup AFTER releasing it;
- Joken.Plug should, perhaps, be on another library with a bigger focus on a more complete authentication library (and perhaps even authorization);

victorolinasc commented:
Let's start 2.0-rc0.
